docs/assignment3.md

```diff
@@ -1 +1,577 @@
+| **Constraints**        | The current code assumes `<action> <item> <preposition> <item>`, which is also like simple commands in natural language. E.g. "give x to y", "drop x", "shout z". If you want to change the command syntax to allow different input, for example more natural language e.g. "I want to give the wallet in my backpack to the guard that is standing at the entrance", this would not be easy. Natural language processing is already a big task on itself, but this would also require a rewrite of both the Interpreter class and the respective methods. |
+
+| ID                     | DP3 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Chain-of-Responsibility Pattern |
+| **Problem**            | We have now defined a way of sending commands to the server in a set language. These different commands all have different actions and need to retrieve or alter different parts of the code. The problem now is that you would not want to couple one player client request to a single class of the server. Coupling one request to a single class is very inflexible, rewriting the handler of the request needs code changes on both ends and makes it impossible to support multiple receivers. |
+| **Solution**           | We need to define a 'chain-of-responsibility'. This allows for one request to be always sent to the 'head' of the chain, and then find its way through to the correct handler. If at the end of the chain the request did not find a way to get handled, we also should to implement a kind of 'safety net' to catch it. |
+| **Intended use**       | When a player sends a command request to the server, it will go through the chain-of-responsibility. Depending on run-time conditions, the correct handler will be chosen. Let's take for example the `shout` command, the arguments of the command should be sent by the server to all other receivers as a broadcasted message. The Player class will receive the command, as it is the head, and forward it to the PlayerThread. It is not their responsibility to handle it, so it will keep forwarding it to the server. The server can then call the respective method to handle this request. <br> Another example is when the player joins the realm for the first time. It is then the PlayerThread's responsibility to receive the request and handle the receiving of the player's username. <br>When an incorrect command is sent over the chain, the request will end up in the last part, in our case the server. The server has a handler for incorrect requests and will reply with an error message. |
+| **Constraints**        | N/A |
+
+| ID                     | DP4 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Decorator Pattern |
```

Comment:
In the decorator design pattern, the decorators should wrap an existing object, implementing its core interface and providing more functionalities. So that the caller is not aware of this additional capabilities.